src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -60659,5 +60681,1008 @@ public void TestMeetsAgree()
+    {
+        if (!t2.HasValue)
+            _ = t2.Value; // 2
+        else
+            _ = t2.Value;
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (8,17): warning CS8629: Nullable value type may be null.
+                //             _ = t1.Value; // 1
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "t1").WithLocation(8, 17),
+                // (13,17): warning CS8629: Nullable value type may be null.
+                //             _ = t2.Value; // 2
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "t2").WithLocation(13, 17));
+        }
+
+        [Fact]
+        public void NullableT_13()
+        {
+            var source =
+@"class C<T> where T : struct
+{
+    static void F1(T? t1)
+    {
+        if (t1 != null)
+            _ = t1.Value;
+        else
+            _ = t1.Value; // 1
+    }
+    static void F2(T? t2)
+    {
+        if (t2 == null)
+            _ = t2.Value; // 2
+        else
+            _ = t2.Value;
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (8,17): warning CS8629: Nullable value type may be null.
+                //             _ = t1.Value; // 1
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "t1").WithLocation(8, 17),
+                // (13,17): warning CS8629: Nullable value type may be null.
+                //             _ = t2.Value; // 2
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "t2").WithLocation(13, 17));
+        }
+
+        [Fact]
+        public void NullableT_14()
+        {
+            var source =
+@"class C<T> where T : struct
+{
+    static void F1(T? t1)
+    {
+        if (null != t1)
+            _ = (T)t1;
+        else
+            _ = (T)t1; // 1
+    }
+    static void F2(T? t2)
+    {
+        if (null == t2)
+            _ = (object)t2; // 2
+        else
+            _ = (object)t2;
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (8,17): warning CS8629: Nullable value type may be null.
+                //             _ = (T)t1; // 1
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "(T)t1").WithLocation(8, 17),
+                // (13,17): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //             _ = (object)t2; // 2
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(object)t2").WithLocation(13, 17));
+        }
+
+        [Fact]
+        public void NullableT_15()
+        {
+            var source =
+@"class C<T> where T : struct
+{
+    static void F1(T? t1)
+    {
+        if (t1.HasValue)
+        {
+            if (t1.HasValue) _ = t1.Value;
```

Comment:
Shouldn't we have a warning here for something along the lines of "this will never be false"?

Reply:
Logged [https://github.com/dotnet/roslyn/issues/31516](https://github.com/dotnet/roslyn/issues/31516).

---
In reply to: [238456621](https://github.com/dotnet/roslyn/pull/31499#discussion_r238456621) [](ancestors = 238456621)